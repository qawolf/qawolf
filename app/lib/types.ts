// GraphQL Types
export type AuthenticatedUser = {
  access_token: string;
  user: User;
};

export type DeploymentProvider = "netlify" | "vercel";

export type Editor = {
  helpers: string;
  run?: Run | null;
  test: Test;
};

export type Env = { [name: string]: string };

export type Environment = {
  id: string;
  name: string;
};

export type EnvironmentVariable = {
  created_at: string;
  environment_id: string;
  id: string;
  name: string;
  value: string;
};

export type GitHubBranch = {
  is_default: boolean;
  name: string;
};

export type Group = {
  id: string;
  name: string;
  team_id: string;
};

export type Integration = {
  github_repo_name: string;
  id: string;
  slack_channel: string | null;
  team_name: string | null;
  type: "github" | "github_sync" | "slack";
};

export type Invite = {
  email: string;
  id: string;
  team_id: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type Onboarding = {
  has_added_trigger_to_test: boolean;
  has_completed_tutorial: boolean;
  has_created_test: boolean;
  has_invited_user: boolean;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  code: string;
  completed_at: string | null;
  created_at: string;
  current_line: number | null;
  environment_id: string | null;
  id: string;
  logs_url: string | null;
  started_at: string | null;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  video_url: string | null;
};

export type Runner = {
  api_key: string | null;
  ws_url: string | null;
};

export type ShortSuite = {
  branch: string | null;
  created_at: string;
  environment_id: string | null;
  environment_name: string | null;
  environment_variables: string | null;
  id: string;
  team_id: string;
  trigger: ShortTrigger | null;
};

export type StatusCounts = {
  created: number;
  fail: number;
  pass: number;
};

export type Suite = ShortSuite & {
  runs: SuiteRun[];
};

export type SuiteRun = {
  completed_at: string | null;
  created_at: string;
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  started_at: string | null;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type SuiteSummary = ShortSuite & {
  status_counts: StatusCounts;
};

export type ShortTeam = {
  id: string;
  name: string;
};

export type Team = ShortTeam & {
  alert_integration_id: string | null;
  alert_only_on_failure: boolean;
  api_key: string;
  git_sync_integration_id: string | null;
  helpers: string;
  inbox: string;
  is_email_alert_enabled: boolean;
  is_enabled: boolean;
  next_trigger_id: string;
  plan: "business" | "custom" | "free";
  renewed_at: string;
};

export type TeamWithUsers = Team & {
  invites: Invite[];
  users: User[];
};

export type ShortTest = {
  group_id: string | null;
  id: string;
  name: string | null;
  path: string | null;
};

export type Test = ShortTest & {
  code: string;
  deleted_at: string | null;
  is_enabled: boolean;
  team_id: string;
  updated_at: string;
  url?: string;
};

export type TestHistoryRun = {
  created_at: string;
  id: string;
  started_at: string;
  status: RunStatus;
};

export type TestSummaryRun = {
  created_at: string;
  gif_url: string | null;
  id: string;
  status: RunStatus;
};

export type TestSummary = {
  gif_url: string | null;
  last_runs: TestSummaryRun[];
  test_id: string;
};

export type ShortTrigger = {
  color: string;
  deployment_provider: DeploymentProvider | null;
  id: string;
  name: string;
  repeat_minutes: number | null;
};

export type Trigger = ShortTrigger & {
  created_at: string;
  deployment_branches: string | null;
  deployment_environment: string | null;
  deployment_integration_id: string | null;
  environment_id: string | null;
  next_at: string | null;
};

export type User = {
  avatar_url: string | null;
  email: string;
  id: string;
  intercom_hash?: string;
  onboarded_at: string | null;
  teams: ShortTeam[];
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Component Types
export type AuthMode = "logIn" | "signUp";

export type CreateCode = {
  code: string;
  test_id: string | null;
};

export type Modal =
  | "confirmBack"
  | "createTest"
  | "editTestsGroup"
  | "environments"
  | "deleteGroup"
  | "deleteTests"
  | "triggers";

export type MutableListArgs = {
  callback: () => void;
  fields?: MutableListFields;
  name: string;
};

export type MutableListFunction = (args: MutableListArgs) => void;

export type MutableListFields = {
  id: string;
  name: string;
};

export type MutableListType = "environment" | "group";

export type NavigationOption = "code" | "logs" | "helpers";

export type NavigationType = "dark" | "light";

export type SelectedTest = {
  group_id?: string;
  id: string;
  name: string;
};

export type Side = "left" | "right";

export type TestTriggers = {
  group_id: string | null;
  test_id: string;
  trigger_ids: string[];
};

export type TriggerFields = {
  deployment_branches: string | null;
  deployment_environment: string | null;
  deployment_integration_id: string | null;
  deployment_provider: DeploymentProvider | null;
  environment_id: string | null;
  name: string;
  repeat_minutes: number | null;
};

export type ValueProp = {
  detail: string;
  message: string;
};

export type Wolf = {
  name: string;
  number: number;
  variant: string;
};

// Runner Types
export type CodeUpdate = {
  code: string;
  generated?: boolean;
  test_id: string;
};

export interface ElementChooserValue {
  isActive: boolean;
  isFillable?: boolean;
  selectors?: string[];
  text?: string;
}

export type Log = {
  message: string;
  severity: "info" | "warning" | "error";
  timestamp: string;
};

export type RunOptions = {
  code: string;
  code_to_run?: string;
  end_line?: number;
  env: Env;
  helpers?: string;
  restart: boolean;
  run_id?: string;
  start_line?: number;
  test_id?: string;
};

export type RunProgress = {
  code: string;
  completed_at: string | null;
  current_line: number | null;
  run_id?: string;
  start_line?: number;
  status: RunStatus;
};

// State Types
type SignUp = {
  inviteId?: string | null;
  redirectUri?: string | null;
};

type ModalState = {
  group?: MutableListFields;
  name: Modal | null;
  testIds?: string[];
  tests?: SelectedTest[];
};

export type State = {
  branch: string | null;
  dashboardUri: string | null;
  editorSidebarWidth: number;
  email: string | null;
  environmentId: string | null;
  isSubscribed: boolean;
  modal: ModalState;
  run: RunOptions | null;
  signUp: SignUp;
  teamId: string | null;
  toast: {
    error?: boolean;
    expiresIn?: number;
    message: string;
  } | null;
};
