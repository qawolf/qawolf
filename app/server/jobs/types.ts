export const JOB_TYPES = [
  "checkPending",
  "deleteRunners",
  "deployRunners",
  "orchestrateGroups",
  "orchestrateRunners",
  "restartRunners",
] as const;

export type Job = typeof JOB_TYPES[number];
