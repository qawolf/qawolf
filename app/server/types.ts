import { Transaction } from "knex";

import { Logger } from "./Logger";

// Azure types
export type SaveArtifacts = {
  gifUrl: string | null;
  logsUrl: string;
  videoUrl: string | null;
};

// Context type
export type Context = {
  api_key: string | null;
  ip: string | null;
  logger: Logger;
  teams: Team[] | null;
  user: User | null;
};

// GitHub types
export type GitHubEmail = {
  email: string;
  primary: boolean;
  verified: boolean;
  visibility?: string | null;
};

export type GitHubFields = {
  avatar_url: string;
  email: string;
  github_id: number;
  github_login: string;
  name: string | null;
};

export type GitHubUser = {
  avatar_url: string;
  email: string;
  id: number;
  login: string;
  name: string;
};

// Slack types
export type SlackWebhook = {
  settings_url: string;
  slack_channel: string;
  slack_team: string;
  webhook_url: string;
};

// Model types
export type ApiKey = {
  created_at: string;
  id: string;
  last_used_at: string | null;
  name: string;
  team_id: string;
  token_digest: string;
  token_end: string;
  updated_at?: string;
};

type WolfFields = {
  wolf_name?: string | null;
  wolf_number?: number | null;
  wolf_variant?: string | null;
};

export type CreateUserWithEmail = WolfFields & {
  email: string;
  login_code: string;
};

export type CreateUserWithGitHub = GitHubFields & WolfFields;

export type DeploymentEnvironment = "preview" | "production";

export type EnvironmentVariable = {
  created_at?: string;
  group_id: string;
  id: string;
  is_system?: boolean;
  name: string;
  team_id: string;
  value: string;
  updated_at?: string;
};

export type FormattedVariables = { [name: string]: string };

export type GitHubCommitStatus = {
  completed_at: string | null;
  context: string;
  created_at?: string;
  deployment_url: string;
  github_installation_id: number;
  group_id: string;
  id: string;
  owner: string;
  repo: string;
  sha: string;
  suite_id: string;
  updated_at?: string;
};

export type Invite = {
  accepted_at: string | null;
  created_at?: string;
  creator_id: string;
  email: string;
  expires_at: string;
  id: string;
  team_id: string;
  updated_at?: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

export type Group = {
  alert_integration_id: string | null;
  created_at?: string;
  creator_id: string;
  deleted_at: string | null;
  deployment_branches?: string | null;
  deployment_environment?: DeploymentEnvironment | null;
  deployment_integration_id: string | null;
  id: string;
  is_default: boolean;
  is_email_enabled: boolean;
  name: string;
  next_at: string | null;
  repeat_minutes: number | null;
  team_id: string;
  updated_at?: string;
};

export type GroupTest = {
  group_id: string;
  id: string;
  test_id: string;
};

export type IntegrationType = "github" | "slack";

export type Integration = {
  created_at?: string;
  github_installation_id?: number | null;
  github_repo_id?: number | null;
  github_repo_name?: string | null;
  id: string;
  settings_url: string | null;
  slack_channel?: string | null;
  team_id: string;
  team_name: string | null;
  type: IntegrationType;
  updated_at?: string;
  webhook_url?: string | null;
};

export type ModelOptions = {
  trx?: Transaction;
  logger: Logger;
};

export type RunStatus = "created" | "fail" | "pass";

export type Run = {
  created_at?: string;
  code: string;
  completed_at?: string;
  current_line: number | null;
  id: string;
  started_at?: string;
  status: RunStatus;
  suite_id: string | null;
  test_id: string;
  updated_at?: string;
};

export type Runner = {
  api_key?: string | null;
  created_at: string;
  deleted_at?: string | null;
  deployed_at?: string | null;
  health_checked_at?: string | null;
  id: string;
  location: string;
  ready_at?: string | null;
  restarted_at?: string | null;
  run_id?: string | null;
  session_expires_at?: string | null;
  test_id?: string | null;
  updated_at?: string;
};

export type RunnerLocation = {
  buffer: number;
  latitude: number;
  longitude: number;
  reserved: number;
};

export type RunnerLocations = Record<string, RunnerLocation>;

export type Suite = {
  alert_sent_at?: string;
  created_at: string;
  creator_id: string | null;
  environment_variables: string | null;
  group_id: string;
  id: string;
  team_id: string;
  updated_at?: string;
};

export type SuiteRun = {
  gif_url: string | null;
  id: string;
  is_test_deleted: boolean;
  status: RunStatus;
  test_id: string;
  test_name: string;
};

export type TeamPlan = "free" | "business";

export type Team = {
  created_at?: string;
  deleted_at?: string | null;
  id: string;
  is_enabled: boolean;
  name: string;
  plan: TeamPlan;
  renewed_at: string | null;
  stripe_customer_id: string | null;
  stripe_subscription_id: string | null;
  updated_at?: string;
};

export type TeamUserRole = "admin";

export type TeamUser = {
  id: string;
  role: TeamUserRole;
  team_id: string;
  user_id: string;
};

export type Test = {
  created_at: string;
  creator_id: string;
  code: string;
  deleted_at: string | null;
  id: string;
  is_enabled: boolean;
  name: string;
  runner_locations?: string | null;
  runner_requested_at?: string | null;
  team_id: string;
  url: string;
  updated_at: string;
  version: number;
};

export type TestUpdate = {
  code: string;
  generated?: boolean;
  id: string;
  version: number;
};

export type User = {
  avatar_url: string | null;
  created_at?: string;
  email: string;
  github_id: number | null;
  github_login: string | null;
  id: string;
  is_enabled?: boolean;
  login_code_digest: string | null;
  login_code_expires_at: string | null;
  name: string | null;
  onboarded_at: string | null;
  updated_at?: string;
  wolf_name: string;
  wolf_number: number;
  wolf_variant: string;
};

// Mutaton types
export type AuthenticatedUser = {
  access_token: string;
  user: CurrentUser;
};

export type CreateApiKeyMutation = {
  name: string;
  team_id: string;
};

export type CreateGitHubIntegrationsMutation = {
  installation_id: number;
  team_id: string;
};

export type CreateEnvironmentVariableMutation = {
  group_id: string;
  name: string;
  value: string;
};

export type CreateInviteMutation = {
  emails: string[];
  team_id: string;
};

export type CreateSlackIntegrationMutation = {
  group_id: string;
  redirect_uri: string;
  slack_code: string;
};

export type CreateSuiteMutation = {
  group_id: string;
  test_ids?: string[] | null;
};

export type CreateTestMutation = {
  group_id: string | null;
  url: string;
};

export type CreateUrlMutation = {
  redirect_uri: string;
};

export type DeleteGroup = {
  default_group_id: string;
  id: string;
};

export type DeleteTestsMutation = {
  ids: string[];
};

export type InstrumentTestRunMutation = {
  status: RunStatus;
  test_id: string;
};

export type JoinMailingListMutation = {
  email: string;
};

export type RunnerRun = {
  artifacts: SaveArtifacts;
  code: string;
  env: string;
  id: string;
  test_id: string;
  version: number;
};

export type RunTestStatus = {
  message: string;
  pass: boolean;
};

export type SendLoginCodeMutation = {
  email: string;
  invite_id?: string | null;
};

export type SendLoginCode = {
  email: string;
};

export type SignInWithEmailMutation = {
  email: string;
  login_code: string;
};

export type SignInWithGitHubMutation = {
  github_code: string;
  github_state: string;
  invite_id?: string | null;
};

export type UpdateGroupMutation = {
  deployment_branches?: string | null;
  deployment_environment?: DeploymentEnvironment | null;
  deployment_integration_id?: string | null;
  id: string;
  is_email_enabled?: boolean;
  name?: string;
  alert_integration_id?: string | null;
  repeat_minutes?: number | null;
};

export type UpdateGroupTestsMutation = {
  add_group_id: string | null;
  remove_group_id: string | null;
  test_ids: string[];
};

export type UpdateRunMutation = {
  current_line: number | null;
  id: string;
  status: RunStatus;
};

export type UpdateRunnerMutation = {
  id: string;
  is_healthy?: boolean;
  is_ready?: boolean;
};

export type UpdateTeamMutation = {
  id: string;
  name: string;
};

export type UpdateTestMutation = {
  code?: string;
  id: string;
  is_enabled?: boolean;
  name?: string;
  version?: number;
};

export type UpdateUserMutation = {
  onboarded_at: string;
};

export type UseTestSessionMutation = {
  test_id: string;
};

// Query types
export type CurrentUser = User & {
  teams: Team[];
};

export type GroupIdQuery = {
  group_id: string;
};

export type IdQuery = {
  id: string;
};

export type RunResult = Run & {
  logs_url: string | null;
  video_url: string | null;
};

export type RunnerResult = {
  api_key?: string;
  ws_url?: string;
};

export type TeamIdQuery = {
  team_id: string;
};

export type TestQuery = {
  id?: string;
  run_id?: string;
};

export type TestResult = {
  run: RunResult | null;
  test: Test;
};

export type TestSummary = {
  gif_url: string | null;
  last_runs: SuiteRun[];
};

export type Wolf = {
  name: string;
  number: string;
  variant: string;
};
