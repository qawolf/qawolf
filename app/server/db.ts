import cuid from "cuid";
import knex from "knex";
import path from "path";
import { Client } from "pg";

import environment from "./environment";

const TEST_DATABASE_NAME = `test_${cuid()}`;

export const createTestDb = async (): Promise<void> => {
  const client = new Client();
  await client.connect();
  await client.query(`CREATE DATABASE ${TEST_DATABASE_NAME}`);
  await client.end();
};

let databaseReadyPromise = Promise.resolve();
if (environment.NODE_ENV === "test") {
  databaseReadyPromise = createTestDb();
}

export const db = knex({
  client: "pg",
  connection: async () => {
    await databaseReadyPromise;

    if (environment.NODE_ENV === "test") {
      return {
        database: TEST_DATABASE_NAME,
        password: process.env.PGPASSWORD,
        user: process.env.PGUSER,
      };
    }

    const connection: knex.PgConnectionConfig = {
      connectionString: environment.DATABASE_URL,
    };
    if (environment.DATABASE_SSL) connection.ssl = environment.DATABASE_SSL;

    return connection;
  },
});

export const dropTestDb = async (): Promise<void> => {
  await db.destroy();
  const client = new Client();
  await client.connect();
  await client.query(`DROP DATABASE ${TEST_DATABASE_NAME}`);
  await client.end();
};

export const migrateDb = async (): Promise<void> => {
  await db.migrate.latest({ directory: path.join(__dirname, "migrations") });
};
