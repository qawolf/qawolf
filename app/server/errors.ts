import {
  ApolloError,
  AuthenticationError as ApolloAuthenticationError,
} from "apollo-server-micro";

// These errors are forwarded to the client / user
export const CLIENT_ERROR_CODES = ["CLIENT", "UNAUTHENTICATED"];

export const AuthenticationError = ApolloAuthenticationError;

export class ClientError extends ApolloError {
  constructor(message: string) {
    super(message, "CLIENT");

    Object.defineProperty(this, "name", { value: "ClientError" });
  }
}
