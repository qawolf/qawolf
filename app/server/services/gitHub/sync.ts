import { RestEndpointMethodTypes } from "@octokit/rest";
import { camelCase } from "lodash";

// import { connectDb } from "../../db";
// import { Logger } from "../../Logger";
// import { findTeam } from "../../models/team";
import { findTestsForTeam, updateTest } from "../../models/test";
import { ModelOptions, Team } from "../../types";
import { createOctokitForIntegration, OctokitRepo } from "./app";
import { findDefaultBranch } from "./branch";

export type Tree = RestEndpointMethodTypes["git"]["createTree"]["parameters"]["tree"];

type CreateCommit = {
  branch?: string;
  message: string;
  tree: Tree;
  team: Team;
};

type CreateCommitForTree = OctokitRepo & {
  message: string;
  parents: string[];
  treeSha: string;
};

type CreateTree = OctokitRepo & {
  tree: Tree;
  treeSha: string;
};

type CurrentCommit = {
  sha: string;
  treeSha: string;
};

type FindCurrentCommit = OctokitRepo & { branch: string };

type UpdateRef = OctokitRepo & {
  branch: string;
  sha: string;
};

export const BLOB_MODE = "100644" as const; // blob

export const buildQaWolfTree = async (
  team: Team,
  options: ModelOptions
): Promise<Tree> => {
  const tests = await findTestsForTeam(team.id, options);

  const helperFiles = [
    {
      content: team.helpers,
      mode: BLOB_MODE,
      path: `qawolf/helpers/index.js`,
    },
  ];

  const filteredTests = tests.filter((test) => !test.guide);

  const testFiles = filteredTests.map((test) => {
    return {
      content: test.code,
      mode: BLOB_MODE,
      path: `qawolf/${camelCase(test.name)}.test.js`,
    };
  });

  await Promise.all(
    filteredTests.map((test) => {
      return updateTest(
        { id: test.id, name: null, path: `${camelCase(test.name)}.test.js` },
        options
      );
    })
  );

  return [...testFiles, ...helperFiles];
};

export const createCommitForTree = async (
  { message, octokit, owner, repo, parents, treeSha }: CreateCommitForTree,
  { logger }: ModelOptions
): Promise<string> => {
  const log = logger.prefix("createCommitForTree");

  const { data } = await octokit.git.createCommit({
    message,
    owner,
    parents,
    repo,
    tree: treeSha,
  });
  const sha = data.sha;
  log.debug("commit sha", sha);

  return sha;
};

export const createTree = async (
  { octokit, owner, repo, tree, treeSha }: CreateTree,
  { logger }: ModelOptions
): Promise<string> => {
  const log = logger.prefix("createTree");

  const { data } = await octokit.git.createTree({
    base_tree: treeSha,
    owner,
    repo,
    tree,
  });
  const sha = data.sha;
  log.debug("tree sha", sha);

  return sha;
};

export const findCurrentCommit = async (
  { branch, octokit, owner, repo }: FindCurrentCommit,
  { logger }: ModelOptions
): Promise<CurrentCommit> => {
  const log = logger.prefix("findCurrentCommit");

  const { data: refData } = await octokit.git.getRef({
    owner,
    ref: `heads/${branch}`,
    repo,
  });
  const sha = refData.object.sha;
  log.debug("head sha", sha);

  const { data: commitData } = await octokit.git.getCommit({
    commit_sha: sha,
    owner,
    repo,
  });

  return { sha, treeSha: commitData.tree.sha };
};

export const updateRef = async ({
  branch,
  octokit,
  owner,
  repo,
  sha,
}: UpdateRef): Promise<void> => {
  await octokit.git.updateRef({
    owner,
    ref: `heads/${branch}`,
    repo,
    sha,
  });
};

export const createCommit = async (
  { branch: passedBranch, message, tree, team }: CreateCommit,
  options: ModelOptions
): Promise<void> => {
  const log = options.logger.prefix("createCommit");
  log.debug("team", team.id);

  const fields = await createOctokitForIntegration(
    team.git_sync_integration_id,
    options
  );

  const branch = passedBranch || (await findDefaultBranch(fields, options));
  const currentCommit = await findCurrentCommit({ ...fields, branch }, options);

  const treeSha = await createTree(
    {
      ...fields,
      tree,
      treeSha: currentCommit.treeSha,
    },
    options
  );

  const commitSha = await createCommitForTree(
    {
      ...fields,
      message,
      parents: [currentCommit.sha],
      treeSha,
    },
    options
  );

  await updateRef({ ...fields, branch, sha: commitSha });

  log.debug("success");
};

// (async () => {
//   const options = { db: connectDb(), logger: new Logger() };

//   const team = await findTeam("teamId", options);
//   const tree = await buildQaWolfTree(team, options);

//   createCommit(
//     { message: "initial qawolf commit", team, tree },
//     options
//   );
// })();
