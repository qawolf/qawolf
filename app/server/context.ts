import { NextApiRequest } from "next";

import { AuthenticationError } from "./errors";
import { Logger } from "./Logger";
import { findTeamsForUser } from "./models/team";
import { findUser } from "./models/user";
import { verifyAccessToken } from "./services/access";
import { Context } from "./types";
import { API_KEY_PREFIX } from "./utils";

const formatIp = (ip: string | null): string | null => {
  if (!ip) return null;

  const formattedIp = new URL(`http://${ip}`).hostname;

  return formattedIp;
};

export const context = async ({
  req,
}: {
  req: NextApiRequest;
}): Promise<Context> => {
  const logger = new Logger({ prefix: "graphql" });

  const authToken = req.headers.authorization || "";

  const api_key = authToken.includes(API_KEY_PREFIX) ? authToken : null;
  const ip = formatIp((req.headers["x-forwarded-for"] as string) || null);
  const user_id = authToken.includes(API_KEY_PREFIX)
    ? null
    : verifyAccessToken(authToken);

  if (!user_id) {
    logger.debug("no current user");
    return { api_key, ip, logger, teams: null, user: null };
  }

  const user = await findUser({ id: user_id }, { logger });

  if (user && !user.is_enabled) {
    throw new AuthenticationError(
      "This account is disabled, email us at hello@qawolf.com"
    );
  }

  const teams = user ? await findTeamsForUser(user.id, { logger }) : null;
  const teamIds = teams?.map((team) => team.id);

  logger.debug(`user ${user?.id} teams ${teamIds}`);

  return { api_key, ip, logger, teams, user };
};
