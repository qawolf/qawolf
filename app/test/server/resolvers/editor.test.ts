import {
  buildTestCode,
  buildTreeForCommit,
  commitTestAndHelpers,
  editorResolver,
  findHelpersForEditor,
  saveEditorResolver,
  updateTestAndHelpers,
} from "../../../server/resolvers/editor";
import * as syncService from "../../../server/services/gitHub/sync";
import * as treeService from "../../../server/services/gitHub/tree";
import { Editor, RunResult } from "../../../server/types";
import { prepareTestDb } from "../db";
import {
  buildIntegration,
  buildRun,
  buildSuite,
  buildTeam,
  buildTest,
  buildTrigger,
  buildUser,
  testContext,
} from "../utils";

const files = [
  { path: "qawolf/helpers/index.js", sha: "sha", text: "git helpers" },
  { path: "qawolf/myTest.test.js", sha: "sha", text: "git code" },
];

const run = buildRun({ suite_id: "suiteId" });
const team = buildTeam({
  git_sync_integration_id: "integrationId",
  helpers: "team helpers",
});
const test = buildTest({ code: "test code", path: "qawolf/myTest.test.js" });

const db = prepareTestDb();
const context = { ...testContext, db, teams: [team] };
const options = { db, logger: context.logger };

beforeAll(async () => {
  await db("teams").insert({ ...team, git_sync_integration_id: null });
  await db("users").insert(buildUser({}));

  await db("integrations").insert(buildIntegration({}));
  await db("teams")
    .where({ id: "teamId" })
    .update({ git_sync_integration_id: "integrationId" });

  await db("triggers").insert(buildTrigger({}));
  await db("suites").insert(buildSuite({ helpers: "suite helpers" }));

  await db("tests").insert(test);
  await db("runs").insert(run);
});

describe("buildTestCode", () => {
  it("returns code from git files if possible", () => {
    const code = buildTestCode({ files, test }, options);

    expect(code).toBe("git code");
  });

  it("returns test code if no files", () => {
    const code = buildTestCode({ files: null, test }, options);

    expect(code).toBe("test code");
  });

  it("throws an error if files but test not found", () => {
    expect(() => {
      buildTestCode({ files: [], test }, options);
    }).toThrowError("not found");
  });
});

describe("buildTreeForCommit", () => {
  it("handles renaming the file", () => {
    expect(
      buildTreeForCommit({ path: "qawolf/renamed.test.js", testFile: files[1] })
    ).toMatchSnapshot();
  });

  it("handles updating code and helpers", () => {
    expect(
      buildTreeForCommit({
        code: "// new code",
        helpers: "// new helpers",
        testFile: files[1],
      })
    ).toMatchSnapshot();
  });
});

describe("commitTestAndHelpers", () => {
  beforeEach(() => {
    jest
      .spyOn(treeService, "findFilesForBranch")
      .mockResolvedValue({ files, owner: "qawolf", repo: "repo" });

    jest.spyOn(syncService, "createCommit").mockResolvedValue();
  });

  afterEach(() => jest.clearAllMocks());

  it("updates test path if applicable", async () => {
    const { helpers, test: updatedTest } = await commitTestAndHelpers(
      {
        branch: "feature",
        path: "qawolf/renamed.test.js",
        team,
        test,
      },
      options
    );

    expect(helpers).toBe("git helpers");
    expect(updatedTest.path).toBe("qawolf/renamed.test.js");

    await db("tests").where({ id: test.id }).update({ path: test.path });
  });

  it("throws an error if test or helpers not found", async () => {
    await expect(
      (): Promise<Editor> => {
        return commitTestAndHelpers(
          {
            branch: "feature",
            code: "// new code",
            team,
            test: { ...test, path: "qawolf/oops.test.js" },
          },
          options
        );
      }
    ).rejects.toThrowError("No helpers or test file");
  });
});

describe("editorResolver", () => {
  afterEach(() => jest.clearAllMocks());

  it("returns helpers, run, and test without branch", async () => {
    jest.spyOn(treeService, "findFilesForBranch");

    const { helpers, run, test } = await editorResolver(
      {},
      { branch: null, test_id: "testId" },
      context
    );

    expect(helpers).toBe("team helpers");
    expect(run).toBeNull();
    expect(test).toMatchObject({ code: "test code", id: "testId" });

    expect(treeService.findFilesForBranch).not.toBeCalled();
  });

  it("returns helpers, run, and test for a run", async () => {
    const { helpers, run, test } = await editorResolver(
      {},
      { branch: null, run_id: "runId" },
      context
    );

    expect(helpers).toBe("suite helpers");
    expect(run).toMatchObject({ id: "runId" });
    expect(test).toMatchObject({ code: "test code", id: "testId" });

    expect(treeService.findFilesForBranch).not.toBeCalled();
  });

  it("returns helpers, run, and test with branch", async () => {
    jest
      .spyOn(treeService, "findFilesForBranch")
      .mockResolvedValue({ files, owner: "qawolf", repo: "repo" });

    const { helpers, run, test } = await editorResolver(
      {},
      { branch: "feature", test_id: "testId" },
      context
    );

    expect(helpers).toBe("git helpers");
    expect(run).toBeNull();
    expect(test).toMatchObject({ code: "git code", id: "testId" });

    expect(treeService.findFilesForBranch).toBeCalled();
  });

  it("throws an error if no run or test id passed", async () => {
    await expect(
      async (): Promise<Editor> => {
        return editorResolver({}, {}, context);
      }
    ).rejects.toThrowError("Must provide test_id or run_id");
  });
});

describe("findHelpersForEditor", () => {
  it("returns helpers from suite if run passed", async () => {
    const helpers = await findHelpersForEditor(
      { files: null, run: run as RunResult, team },
      options
    );

    expect(helpers).toBe("suite helpers");
  });

  it("returns helpers from files if files passed", async () => {
    const helpers = await findHelpersForEditor(
      { files, run: null, team },
      options
    );

    expect(helpers).toBe("git helpers");
  });

  it("returns helpers from team otherwise", async () => {
    const helpers = await findHelpersForEditor(
      { files: null, run: null, team },
      options
    );

    expect(helpers).toBe("team helpers");
  });
});

describe("saveEditorResolver", () => {
  it("updates test and helpers without branch", async () => {
    await db("tests")
      .where({ id: "testId" })
      .update({ name: "old name", path: null });

    const { helpers, test: updatedTest } = await saveEditorResolver(
      {},
      {
        code: "new code",
        helpers: "new helpers",
        name: "new name",
        test_id: "testId",
      },
      context
    );

    expect(helpers).toBe("new helpers");
    expect(updatedTest).toMatchObject({ code: "new code", name: "new name" });

    await db("teams").where({ id: "teamId" }).update({ helpers: team.helpers });
    await db("tests")
      .where({ id: "testId" })
      .update({ code: test.code, name: null, path: test.path });
  });
});

describe("updateTestAndHelpers", () => {
  it("updates helpers", async () => {
    const { helpers, test: updatedTest } = await updateTestAndHelpers(
      { helpers: "new helpers", name: null, team, test },
      options
    );

    expect(helpers).toBe("new helpers");
    expect(updatedTest.code).toBe(test.code);
    expect(updatedTest.name).toBe(test.name);

    const updatedTeam = await db("teams").first();
    expect(updatedTeam.helpers).toBe("new helpers");

    await db("teams").where({ id: "teamId" }).update({ helpers: team.helpers });
  });

  it("updates test code and name", async () => {
    await db("tests")
      .where({ id: "testId" })
      .update({ name: "old name", path: null });

    const { helpers, test: updatedTest } = await updateTestAndHelpers(
      { code: "new code", name: "new name", team, test },
      options
    );

    expect(helpers).toBe("team helpers");
    expect(updatedTest).toMatchObject({ code: "new code", name: "new name" });

    await db("tests")
      .where({ id: "testId" })
      .update({ code: test.code, name: null, path: test.path });
  });
});
