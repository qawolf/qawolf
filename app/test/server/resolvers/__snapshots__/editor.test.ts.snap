// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildTreeForCommit handles renaming the file 1`] = `
Object {
  "message": "rename qawolf/myTest.test.js",
  "tree": Array [
    Object {
      "mode": "100644",
      "path": "qawolf/myTest.test.js",
      "sha": null,
    },
    Object {
      "content": "git code",
      "mode": "100644",
      "path": "qawolf/renamed.test.js",
    },
  ],
}
`;

exports[`buildTreeForCommit handles updating code and helpers 1`] = `
Object {
  "message": "update qawolf/myTest.test.js, update qawolf/helpers/index.js",
  "tree": Array [
    Object {
      "content": "// new code",
      "mode": "100644",
      "path": "qawolf/myTest.test.js",
    },
    Object {
      "content": "// new helpers",
      "mode": "100644",
      "path": "qawolf/helpers/index.js",
    },
  ],
}
`;
